### PR TITLE
refactor(error-overlay): unify Pages/App router error overlay source

### PR DESCRIPTION
**Pull request by balazsorban44**
This is a further iteration on #60899 which first moved the Pages error overlay from a different package into the `next` directly. While working on #62942 I noticed that a lot of the logic is shared between the two routers' error overlay implementation.

A lot of type/code duplication can be spared, using the https://github.com/vercel/next.js/blob/feat/unified-app-pages-error-overlay/packages/next/src/client/components/react-dev-overlay/shared.ts file.

Since there is no significant difference between the HMR events between Pages/App, it's easier to maintain a single shared implementation than remembering to sync changes in multiple places.

I also moved the Pages overlay-related changes into the same level where App Router's overlay is defined, so the file structure is easier to traverse/understand.

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 31ea7d26f8da0129bafa70b0961986a3014a0a34

`TURBOPACK=1 pnpm test test/integration/next-image-new/invalid-image-import/test/index.test.ts` (turbopack)
- Missing Import Image Tests > dev mode > should show error

<details>
<summary>Expand output</summary>

  ● Missing Import Image Tests › dev mode › should show error

    expect(received).toBe(expected) // Object.is equality

    - Expected  - 0
    + Received  + 1

      Failed to compile
    + Next.js (14.2.0-canary.4) (turbo)

      24 |       const browser = await webdriver(appPort, '/')
      25 |       expect(await hasRedbox(browser)).toBe(true)
    > 26 |       expect(await getRedboxHeader(browser)).toBe('Failed to compile')
         |                                              ^
      27 |       const source = await getRedboxSource(browser)
      28 |       if (process.env.TURBOPACK) {
      29 |         expect(source).toMatchInlineSnapshot(`

      at Object.toBe (integration/next-image-new/invalid-image-import/test/index.test.ts:26:46)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.7s | N/A |
| buildDurationCached | 7.4s | 6.2s | N/A |
| nodeModulesSize | 198 MB | 198 MB | N/A |
| nextStartRea..uration (ms) | 432ms | 434ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.1 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.56 kB | 6.56 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 542 B | 540 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 528 B | 528 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95 kB | 95 kB | N/A |
| page.js gzip | 3.07 kB | 3.07 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.7 kB | 96.7 kB | ✓ |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | ✓ |
| app-page-tur..prod.js gzip | 92.9 kB | 92.9 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.4 kB | 91.4 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.54 kB | 9.54 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.81 kB | 9.81 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.54 kB | 9.54 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.6 kB | ✓ |
| Overall change | 949 kB | 949 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js feat/unified-app-pages-error-overlay | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 106 kB | 106 kB | ⚠️ +216 B |
| Overall change | 106 kB | 106 kB | ⚠️ +216 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 95c5d8bd5afcfbcae02b53161119af53b4567eea

**Review by kdy1 — APPROVED** [review]

**Comment by kdy1** [comment]
(I reviewed to learn about source code related to error overlay, as I'm working on PACK-2405)

File: packages/next/src/client/components/react-dev-overlay/pages/ErrorBoundary.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import React from 'react'
+import * as React from 'react'
```
Comment (huozhi) [inline packages/next/src/client/components/react-dev-overlay/pages/ErrorBoundary.tsx:1]: React here is used as a type, no need to use `import *`

**Review by huozhi — APPROVED** [review]

**Pull request merged**